presto/dataset.py

```diff
@@ -394,72 +378,220 @@ def __len__(self):
 
-            eo_data[:, :, BANDS.index(presto_val)] = values
-            mask[:, :, IDX_TO_BAND_GROUPS[presto_val]] += ~idx_valid
+        start_month = (inarr.t.values[0].astype("datetime64[M]").astype(int) % 12 + 1) - 1
 
-        y = rearrange(ds[cls.Y].values, "t x y -> (x y) t")
-        # -1 because we index from 0
-        start_month = (ds.t.values[0].astype("datetime64[M]").astype(int) % 12 + 1) - 1
         months = np.ones((num_instances)) * start_month
+        return months
 
-        transformer = Transformer.from_crs(f"EPSG:{epsg_coords}", "EPSG:4326", always_xy=True)
-        lon, lat = transformer.transform(ds.x, ds.y)
+    @staticmethod
+    def _subset_array_temporally(inarr: xr.DataArray) -> xr.DataArray:
+        """
+        Subset the input xarray.DataArray temporally based on `valid_time` attribute.
 
-        latlons = np.stack(
-            [np.repeat(lat, repeats=len(lon)), repeat(lon, "c -> (h c)", h=len(lat))],
-            axis=-1,
-        )
+        Args:
+            inarr (xr.DataArray): Input xarray.DataArray containing EO data.
+
+        Returns:
+            xr.DataArray: Temporally subsetted xarray.DataArray.
+        """
 
-        return eo_data, np.repeat(mask, BAND_EXPANSION, axis=-1), latlons, months, y
+        # Use valid_time attribute to extract the right part of the time series
+        valid_time = pd.to_datetime(inarr.attrs["valid_time"]).replace(day=1)
+        end_time = valid_time + pd.DateOffset(months=5)
+        start_time = valid_time - pd.DateOffset(months=6)
+        inarr = inarr.sel(t=slice(start_time, end_time))
+        num_timesteps = len(inarr.t)
+        assert num_timesteps == 12, "Expected 12 timesteps, only found {}".format(num_timesteps)
+
+        return inarr
+
+    @classmethod
+    def nc_to_arrays(
+        cls, filepath: Path
+    ) -> Tuple[
+        np.ndarray,
+        np.ndarray,
+        np.ndarray,
+        np.ndarray,
+        np.ndarray,
+        np.ndarray,
+        np.ndarray,
+    ]:
+        ds = xr.open_dataset(filepath)
+        epsg = CRS.from_wkt(xr.open_dataset(filepath).crs.attrs["crs_wkt"]).to_epsg()
+
+        if epsg is None:
+            raise ValueError("EPSG code not found in the input file.")
+        inarr = ds.drop("crs").to_array(dim="bands")
+        lon, lat = inarr.y.values, inarr.x.values
+
+        # Temporal subsetting to 12 timesteps
+        inarr = cls._subset_array_temporally(inarr)
+
+        eo_data, mask = cls._extract_eo_data(inarr)
+        latlons = cls._extract_latlons(inarr, epsg)
+        months = cls._extract_months(inarr)
+
+        if cls.Y not in ds:
+            target = np.ones_like(months) * cls._NODATAVALUE
+        else:
+            target = rearrange(inarr.sel(bands=cls.Y).values, "t x y -> (x y) t")
+
+        return (
+            eo_data,
+            np.repeat(mask, BAND_EXPANSION, axis=-1),
+            latlons,
+            months,
+            target,
+            lon,
+            lat,
```

Comment:
Do we need `latlons` and `lon, lat` here? If I understand correctly, `lon == latlons[:, 1]` and `lat == latlons[:, 0]`, which means we don't need `lon, lat`?

Reply:
It's not so simple I'm afraid, I struggled with this a lot. After the meshgrid and flattening of `latlons` it becomes quite hard to get back to original `lon, lat`. 

```
latlons.shape
(2500, 2)
lon.shape
(50,)
```
So no, `lon != latlons[:, 1]`. I've been thinking about easier ways but haven't found them as of yet.

Reply:
hmm okay. I think this is probably (?) easier, especially considering we additionally apply the transformation. The latlons take up quite a bit of RAM, and for large tiles this might become an issue.

Just in case its useful here is some code to go from the flat latlons back to the original ones (but without the transformation):

```python
import numpy as np
from einops import rearrange

org_lat = np.array([1, 2, 3])
org_lon = np.array([4, 5, 6, 7])

def to_flat_latlons(lat, lon):
    lon, lat = np.meshgrid(lon, lat)
    latlons = rearrange(np.stack([lat, lon]), "c x y -> (x y) c")
    return latlons


def from_latlons(latlons):
    x = len(np.unique(latlons[:, 0]))
    y = len(np.unique(latlons[:, 1]))
    latlons = rearrange(latlons, "(x y) c -> c x y", x=x, y=y)
    lats, lons = latlons[0], latlons[1]
    return lats[:, 0], lons[0, ]

output_lat, output_lon = from_latlons(to_flat_latlons(org_lat, org_lon))
assert np.equal(output_lon, org_lon).all()
assert np.equal(output_lat, org_lat).all()
```

I wonder if its worth passing around the original lats and lons and only applying that transformation right before the model ingests the values

Reply:
I hope i understood this well. Could you check if my latest commit address this the way you suggest it? Feel free to suggest to do it differently. Rest of the day, unfortunately I'm away. Will catch up tomorrow morning.